paimon-core/src/main/java/org/apache/paimon/table/ExpireSnapshotsImpl.java

```diff
@@ -215,9 +222,59 @@ public int expireUntil(long earliestId, long endExclusiveId) {
             }
 
             Snapshot snapshot = snapshotManager.snapshot(id);
-            snapshotDeletion.cleanUnusedManifests(snapshot, skippingSet);
-
-            // delete snapshot last
+            if (changelogDecoupled) {
```

Comment:
Maybe you should write changelog first? Image there is a failover here.

Maybe you can add test in `FileStoreExpireTestBase`, to make sure changelog files generated.

Reply:
The snapshot path is deleted after changelog files are generated. The test `org.apache.paimon.operation.ExpireSnapshotsTest#testChangelogOutLivedSnapshot` will verify the changelog metadata is generated.